src/aws_encryption_sdk/keyrings/aws_kms/__init__.py

```diff
@@ -0,0 +1,376 @@
+            decryption_materials=decryption_materials, encrypted_data_keys=encrypted_data_keys
+        )
+
+
+@attr.s
+class _AwsKmsSingleCmkKeyring(Keyring):
+    """AWS KMS keyring that only works with a single AWS KMS CMK.
+
+    This keyring should never be used directly.
+    It should only ever be used internally by :class:`KmsKeyring`.
+
+    .. versionadded:: 1.5.0
+
+    :param str key_id: CMK key ID
+    :param ClientSupplier client_supplier: Client supplier to use when asking for clients
+    :param List[str] grant_tokens: AWS KMS grant tokens to include in requests (optional)
+    """
+
+    _key_id = attr.ib(validator=instance_of(six.string_types))
+    _client_supplier = attr.ib(validator=instance_of(ClientSupplier))
+    _grant_tokens = attr.ib(
+        default=attr.Factory(tuple),
+        validator=(deep_iterable(member_validator=instance_of(six.string_types)), value_is_not_a_string),
+    )
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        trace_info = MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=self._key_id)
+        try:
+            if encryption_materials.data_encryption_key is None:
+                plaintext_key, encrypted_key = _do_aws_kms_generate_data_key(
+                    client_supplier=self._client_supplier,
+                    key_name=self._key_id,
+                    encryption_context=encryption_materials.encryption_context,
+                    algorithm=encryption_materials.algorithm,
+                    grant_tokens=self._grant_tokens,
+                )
+                encryption_materials.add_data_encryption_key(
+                    data_encryption_key=plaintext_key,
+                    keyring_trace=KeyringTrace(wrapping_key=trace_info, flags=_GENERATE_FLAGS),
+                )
+            else:
+                encrypted_key = _do_aws_kms_encrypt(
+                    client_supplier=self._client_supplier,
+                    key_name=self._key_id,
+                    plaintext_data_key=encryption_materials.data_encryption_key,
+                    encryption_context=encryption_materials.encryption_context,
+                    grant_tokens=self._grant_tokens,
+                )
+        except Exception:  # pylint: disable=broad-except
+            # We intentionally WANT to catch all exceptions here
+            message = "Unable to generate or encrypt data key using {}".format(trace_info)
+            _LOGGER.exception(message)
+            raise EncryptKeyError(message)
+
+        encryption_materials.add_encrypted_data_key(
+            encrypted_data_key=encrypted_key, keyring_trace=KeyringTrace(wrapping_key=trace_info, flags=_ENCRYPT_FLAGS)
+        )
+
+        return encryption_materials
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        for edk in encrypted_data_keys:
```

Comment:
There is something about this that I feel sad about.
I feel like there should be a more efficient way to search the EDKs list,
and then _only_ process the matching ones...